test/libjulia/Inliner.cpp

```diff
@@ -0,0 +1,185 @@
+
+	InlinableFunctionFilter filter;
+	filter(ast);
+
+	return boost::algorithm::join(
+		filter.inlinableFunctions() | boost::adaptors::map_keys,
+		","
+	);
+}
+
+string inlineFunctions(string const& _source, bool _julia = true)
+{
+	auto ast = disambiguate(_source, _julia);
+	FunctionalInliner(ast).run();
+	return assembly::AsmPrinter(_julia)(ast);
+}
+}
+
+BOOST_AUTO_TEST_SUITE(IuliaInlinableFunctionFilter)
+
+BOOST_AUTO_TEST_CASE(smoke_test)
+{
+	BOOST_CHECK_EQUAL(inlinableFunctions("{ }"), "");
+}
+
+BOOST_AUTO_TEST_CASE(simple)
+{
+	BOOST_CHECK_EQUAL(inlinableFunctions("{ function f() -> x:u256 { x := 2:u256 } }"), "f");
+	BOOST_CHECK_EQUAL(inlinableFunctions(R"({
+		function g(a:u256) -> b:u256 { b := a }
+		function f() -> x:u256 { x := g(2:u256) }
+	})"), "f,g");
+}
+
+BOOST_AUTO_TEST_CASE(simple_inside_structures)
+{
+	BOOST_CHECK_EQUAL(inlinableFunctions(R"({
+		switch 2:u256
+		case 2:u256 {
+			function g(a:u256) -> b:u256 { b := a }
+			function f() -> x:u256 { x := g(2:u256) }
+		}
+	})"), "f,g");
+	BOOST_CHECK_EQUAL(inlinableFunctions(R"({
+		for {
+			function g(a:u256) -> b:u256 { b := a }
+		} 1:u256 {
+			function f() -> x:u256 { x := g(2:u256) }
+		}
+		{
+			function h() -> y:u256 { y := 2:u256 }
+		}
+	})"), "f,g,h");
+}
+
+BOOST_AUTO_TEST_CASE(negative)
+{
+	BOOST_CHECK_EQUAL(inlinableFunctions("{ function f() -> x:u256 { } }"), "");
+	BOOST_CHECK_EQUAL(inlinableFunctions("{ function f() -> x:u256 { x := 2:u256 {} } }"), "");
+	BOOST_CHECK_EQUAL(inlinableFunctions("{ function f() -> x:u256 { x := f() } }"), "");
+	BOOST_CHECK_EQUAL(inlinableFunctions("{ function f() -> x:u256 { x := x } }"), "");
+	BOOST_CHECK_EQUAL(inlinableFunctions("{ function f() -> x:u256, y:u256 { x := 2:u256 } }"), "");
+}
+
+
+BOOST_AUTO_TEST_SUITE_END()
+
+BOOST_AUTO_TEST_SUITE(IuliaFunctionInliner)
+
+BOOST_AUTO_TEST_CASE(simple)
+{
+	BOOST_CHECK_EQUAL(
+		inlineFunctions("{ function f() -> x:u256 { x := 2:u256 } let y:u256 := f() }"),
+		format("{ function f() -> x:u256 { x := 2:u256 } let y:u256 := 2:u256 }")
+	);
+}
+
+BOOST_AUTO_TEST_CASE(with_args)
+{
+	BOOST_CHECK_EQUAL(
+		inlineFunctions("{ function f(a:u256) -> x:u256 { x := a } let y:u256 := f(7:u256) }"),
+		format("{ function f(a:u256) -> x:u256 { x := a } let y:u256 := 7:u256 }")
+	);
+}
+
+BOOST_AUTO_TEST_CASE(no_inline_with_mload)
+{
+	// Does not inline because mload could be moved out of sequence
+	BOOST_CHECK_EQUAL(
+		inlineFunctions("{ function f(a) -> x { x := a } let y := f(mload(2)) }", false),
+		format("{ function f(a) -> x { x := a } let y := f(mload(2)) }", false)
+	);
+}
+
+BOOST_AUTO_TEST_CASE(no_move_with_side_effects)
+{
+	// The calls to g and h cannot be moved because g and h are not movable. Therefore, the call
+	// to f is not inlined.
+	BOOST_CHECK_EQUAL(
+		inlineFunctions(R"({
+			function f(a, b) -> x { x := add(b, a) }
+			function g() -> y { y := mload(0) mstore(0, 4) }
+			function h() -> z { mstore(0, 4) z := mload(0) }
+			let r := f(g(), h())
```

Comment:
I don't get why `f` cannot be inlined here?

Reply:
`f` cannot be inlined because it might change the evaluation order of `h()` and `g()` and also might evaluate it multiple times or not at all.